### PR TITLE
Benchmark for `strings::repeat_strings` APIs

### PR DESCRIPTION
**Pull request by ttnghia**
This PR implemented benchmarks for the string APIs `repeat_strings`. The benchmark results listed below were generated from the current APIs and also from the same APIs but with some modifications.

Note that this PR includes upstream code from https://github.com/rapidsai/cudf/pull/8561 thus the code from that PR is also listed here as "changed files".

Blocked by https://github.com/rapidsai/cudf/pull/8561.

## Terms:
 * Separate checking: Call for overflow checking separately before actually performing string repeating
 * Integrated checking: Inject overflow checking in the middle of the computation, using the string offsets generated during string repeating

## Benchmark results:
Machine:
```
Run on (36 X 4600 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x18)
  L1 Instruction 32 KiB (x18)
  L2 Unified 1024 KiB (x18)
  L3 Unified 25344 KiB (x1)
Load Average: 1.25, 1.60, 1.89
```
### Without any overflow checking:
```
-------------------------------------------------------------------------------------------------------------
Benchmark                                                   Time             CPU   Iterations UserCounters...
--------------------------------------------------------------------------------------------------------------
repeat_strings_scalar_times/256/16/manual_time          0.043 ms        0.062 ms        17850 bytes_per_second=45.3825M/s
repeat_strings_scalar_times/256/64/manual_time          0.053 ms        0.071 ms        12763 bytes_per_second=156.17M/s
repeat_strings_scalar_times/256/256/manual_time         0.116 ms        0.132 ms         5925 bytes_per_second=296M/s
repeat_strings_scalar_times/1024/16/manual_time         0.040 ms        0.058 ms        17683 bytes_per_second=212.767M/s
repeat_strings_scalar_times/1024/64/manual_time         0.053 ms        0.069 ms        12455 bytes_per_second=636.865M/s
repeat_strings_scalar_times/1024/256/manual_time        0.127 ms        0.141 ms         5406 bytes_per_second=1079.24M/s
repeat_strings_scalar_times/4096/16/manual_time         0.041 ms        0.057 ms        16729 bytes_per_second=808.35M/s
repeat_strings_scalar_times/4096/64/manual_time         0.093 ms        0.107 ms         7181 bytes_per_second=1.43583G/s
repeat_strings_scalar_times/4096/256/manual_time        0.466 ms        0.484 ms         1503 bytes_per_second=1.14725G/s
repeat_strings_scalar_times/16384/16/manual_time        0.057 ms        0.071 ms        11578 bytes_per_second=2.32692G/s
repeat_strings_scalar_times/16384/64/manual_time        0.242 ms        0.260 ms         2840 bytes_per_second=2.21228G/s
repeat_strings_scalar_times/16384/256/manual_time        7.54 ms         7.56 ms           93 bytes_per_second=290.49M/s
repeat_strings_scalar_times/65536/16/manual_time        0.109 ms        0.127 ms         5252 bytes_per_second=4.84485G/s
repeat_strings_scalar_times/65536/64/manual_time        0.816 ms        0.834 ms          849 bytes_per_second=2.62289G/s
repeat_strings_scalar_times/65536/256/manual_time        36.8 ms         36.8 ms           19 bytes_per_second=238.362M/s
repeat_strings_scalar_times/262144/16/manual_time       0.344 ms        0.364 ms         2091 bytes_per_second=6.17523G/s
repeat_strings_scalar_times/262144/64/manual_time        3.12 ms         3.14 ms          223 bytes_per_second=2.73715G/s
repeat_strings_scalar_times/262144/256/manual_time        143 ms          143 ms            5 bytes_per_second=245.057M/s
repeat_strings_column_times/256/16/manual_time          0.068 ms        0.085 ms         8440 bytes_per_second=43.036M/s
repeat_strings_column_times/256/64/manual_time          0.202 ms        0.219 ms         3385 bytes_per_second=45.7583M/s
repeat_strings_column_times/256/256/manual_time         0.820 ms        0.836 ms          784 bytes_per_second=43.1525M/s
repeat_strings_column_times/1024/16/manual_time         0.124 ms        0.142 ms         5174 bytes_per_second=100.509M/s
repeat_strings_column_times/1024/64/manual_time         0.382 ms        0.399 ms         1835 bytes_per_second=98.544M/s
repeat_strings_column_times/1024/256/manual_time         1.65 ms         1.66 ms          398 bytes_per_second=85.3484M/s
repeat_strings_column_times/4096/16/manual_time         0.124 ms        0.141 ms         5279 bytes_per_second=394.454M/s
repeat_strings_column_times/4096/64/manual_time         0.400 ms        0.415 ms         1748 bytes_per_second=380.972M/s
repeat_strings_column_times/4096/256/manual_time         1.66 ms         1.67 ms          416 bytes_per_second=339.81M/s
repeat_strings_column_times/16384/16/manual_time        0.135 ms        0.150 ms         5092 bytes_per_second=1.43102G/s
repeat_strings_column_times/16384/64/manual_time        0.406 ms        0.423 ms         1721 bytes_per_second=1.47128G/s
repeat_strings_column_times/16384/256/manual_time        1.67 ms         1.69 ms          416 bytes_per_second=1.3138G/s
repeat_strings_column_times/65536/16/manual_time        0.415 ms        0.433 ms         1679 bytes_per_second=1.86606G/s
repeat_strings_column_times/65536/64/manual_time         2.51 ms         2.53 ms          277 bytes_per_second=972.971M/s
repeat_strings_column_times/65536/256/manual_time        32.8 ms         32.8 ms           21 bytes_per_second=274.627M/s
repeat_strings_column_times/262144/16/manual_time        3.78 ms         3.80 ms          186 bytes_per_second=839.675M/s
repeat_strings_column_times/262144/64/manual_time        46.1 ms         46.1 ms           15 bytes_per_second=211.801M/s
repeat_strings_column_times/262144/256/manual_time        329 ms          329 ms            2 bytes_per_second=109.597M/s
```

### With separate overflow checking:
```
-------------------------------------------------------------------------------------------------------------
Benchmark                                                   Time             CPU   Iterations UserCounters...
-------------------------------------------------------------------------------------------------------------
repeat_strings_scalar_times/256/16/manual_time          0.052 ms        0.070 ms        12103 bytes_per_second=37.368M/s
repeat_strings_scalar_times/256/64/manual_time          0.065 ms        0.082 ms        11249 bytes_per_second=127.858M/s
repeat_strings_scalar_times/256/256/manual_time         0.129 ms        0.145 ms         5006 bytes_per_second=267.697M/s
repeat_strings_scalar_times/1024/16/manual_time         0.051 ms        0.069 ms        12464 bytes_per_second=165.709M/s
repeat_strings_scalar_times/1024/64/manual_time         0.066 ms        0.082 ms         8230 bytes_per_second=511.448M/s
repeat_strings_scalar_times/1024/256/manual_time        0.164 ms        0.182 ms         4737 bytes_per_second=834.513M/s
repeat_strings_scalar_times/4096/16/manual_time         0.061 ms        0.078 ms         9498 bytes_per_second=542.576M/s
repeat_strings_scalar_times/4096/64/manual_time         0.114 ms        0.129 ms         5767 bytes_per_second=1.16889G/s
repeat_strings_scalar_times/4096/256/manual_time        0.509 ms        0.527 ms         1367 bytes_per_second=1075.86M/s
repeat_strings_scalar_times/16384/16/manual_time        0.072 ms        0.086 ms         8897 bytes_per_second=1.82313G/s
repeat_strings_scalar_times/16384/64/manual_time        0.260 ms        0.279 ms         2690 bytes_per_second=2.06074G/s
repeat_strings_scalar_times/16384/256/manual_time        9.50 ms         9.52 ms           77 bytes_per_second=230.482M/s
repeat_strings_scalar_times/65536/16/manual_time        0.125 ms        0.144 ms         5277 bytes_per_second=4.22483G/s
repeat_strings_scalar_times/65536/64/manual_time        0.866 ms        0.888 ms          767 bytes_per_second=2.47143G/s
repeat_strings_scalar_times/65536/256/manual_time        37.5 ms         37.5 ms           19 bytes_per_second=233.585M/s
repeat_strings_scalar_times/262144/16/manual_time       0.333 ms        0.352 ms         2075 bytes_per_second=6.36716G/s
repeat_strings_scalar_times/262144/64/manual_time        3.13 ms         3.15 ms          224 bytes_per_second=2.73003G/s
repeat_strings_scalar_times/262144/256/manual_time        143 ms          143 ms            5 bytes_per_second=244.985M/s
repeat_strings_column_times/256/16/manual_time          0.093 ms        0.111 ms         6545 bytes_per_second=31.3167M/s
repeat_strings_column_times/256/64/manual_time          0.223 ms        0.240 ms         3031 bytes_per_second=41.4737M/s
repeat_strings_column_times/256/256/manual_time         0.839 ms        0.855 ms          823 bytes_per_second=42.184M/s
repeat_strings_column_times/1024/16/manual_time         0.146 ms        0.164 ms         4211 bytes_per_second=85.1519M/s
repeat_strings_column_times/1024/64/manual_time         0.402 ms        0.419 ms         1747 bytes_per_second=93.7183M/s
repeat_strings_column_times/1024/256/manual_time         1.68 ms         1.70 ms          420 bytes_per_second=83.6342M/s
repeat_strings_column_times/4096/16/manual_time         0.156 ms        0.173 ms         4586 bytes_per_second=314.36M/s
repeat_strings_column_times/4096/64/manual_time         0.423 ms        0.438 ms         1636 bytes_per_second=360.101M/s
repeat_strings_column_times/4096/256/manual_time         1.67 ms         1.69 ms          413 bytes_per_second=337.363M/s
repeat_strings_column_times/16384/16/manual_time        0.157 ms        0.171 ms         4178 bytes_per_second=1.23136G/s
repeat_strings_column_times/16384/64/manual_time        0.428 ms        0.444 ms         1628 bytes_per_second=1.39523G/s
repeat_strings_column_times/16384/256/manual_time        1.70 ms         1.72 ms          411 bytes_per_second=1.29319G/s
repeat_strings_column_times/65536/16/manual_time        0.420 ms        0.438 ms         1577 bytes_per_second=1.84401G/s
repeat_strings_column_times/65536/64/manual_time         2.70 ms         2.72 ms          259 bytes_per_second=905.829M/s
repeat_strings_column_times/65536/256/manual_time        33.3 ms         33.3 ms           21 bytes_per_second=270.614M/s
repeat_strings_column_times/262144/16/manual_time        4.15 ms         4.17 ms          167 bytes_per_second=764.591M/s
repeat_strings_column_times/262144/64/manual_time        46.6 ms         46.6 ms           15 bytes_per_second=209.284M/s
repeat_strings_column_times/262144/256/manual_time        329 ms          329 ms            2 bytes_per_second=109.643M/s
``` 

### With integrated overflow checking:
```
-------------------------------------------------------------------------------------------------------------
Benchmark                                                   Time             CPU   Iterations UserCounters...
-------------------------------------------------------------------------------------------------------------
repeat_strings_scalar_times/256/16/manual_time          0.046 ms        0.064 ms        11316 bytes_per_second=42.1088M/s
repeat_strings_scalar_times/256/64/manual_time          0.061 ms        0.079 ms        10387 bytes_per_second=134.546M/s
repeat_strings_scalar_times/256/256/manual_time         0.130 ms        0.146 ms         5531 bytes_per_second=265.39M/s
repeat_strings_scalar_times/1024/16/manual_time         0.059 ms        0.078 ms         8486 bytes_per_second=143.562M/s
repeat_strings_scalar_times/1024/64/manual_time         0.064 ms        0.081 ms         8424 bytes_per_second=525.084M/s
repeat_strings_scalar_times/1024/256/manual_time        0.152 ms        0.171 ms         4078 bytes_per_second=896.389M/s
repeat_strings_scalar_times/4096/16/manual_time         0.060 ms        0.078 ms        10393 bytes_per_second=554.57M/s
repeat_strings_scalar_times/4096/64/manual_time         0.118 ms        0.133 ms         5616 bytes_per_second=1.12776G/s
repeat_strings_scalar_times/4096/256/manual_time        0.496 ms        0.514 ms         1369 bytes_per_second=1104.09M/s
repeat_strings_scalar_times/16384/16/manual_time        0.073 ms        0.089 ms         8857 bytes_per_second=1.80611G/s
repeat_strings_scalar_times/16384/64/manual_time        0.265 ms        0.286 ms         2676 bytes_per_second=2.02063G/s
repeat_strings_scalar_times/16384/256/manual_time        8.38 ms         8.40 ms           70 bytes_per_second=261.41M/s
repeat_strings_scalar_times/65536/16/manual_time        0.137 ms        0.159 ms         4912 bytes_per_second=3.86253G/s
repeat_strings_scalar_times/65536/64/manual_time        0.827 ms        0.845 ms          842 bytes_per_second=2.59047G/s
repeat_strings_scalar_times/65536/256/manual_time        36.5 ms         36.5 ms           19 bytes_per_second=240.245M/s
repeat_strings_scalar_times/262144/16/manual_time       0.329 ms        0.348 ms         2093 bytes_per_second=6.44757G/s
repeat_strings_scalar_times/262144/64/manual_time        3.12 ms         3.14 ms          225 bytes_per_second=2.74115G/s
repeat_strings_scalar_times/262144/256/manual_time        143 ms          143 ms            5 bytes_per_second=245.422M/s
repeat_strings_column_times/256/16/manual_time          0.076 ms        0.094 ms         8701 bytes_per_second=38.2733M/s
repeat_strings_column_times/256/64/manual_time          0.210 ms        0.228 ms         3299 bytes_per_second=43.8627M/s
repeat_strings_column_times/256/256/manual_time         0.831 ms        0.847 ms          834 bytes_per_second=42.5716M/s
repeat_strings_column_times/1024/16/manual_time         0.129 ms        0.147 ms         5279 bytes_per_second=96.2961M/s
repeat_strings_column_times/1024/64/manual_time         0.388 ms        0.405 ms         1785 bytes_per_second=97.1464M/s
repeat_strings_column_times/1024/256/manual_time         1.65 ms         1.67 ms          421 bytes_per_second=85.0178M/s
repeat_strings_column_times/4096/16/manual_time         0.132 ms        0.149 ms         5124 bytes_per_second=369.813M/s
repeat_strings_column_times/4096/64/manual_time         0.408 ms        0.423 ms         1711 bytes_per_second=373.09M/s
repeat_strings_column_times/4096/256/manual_time         1.66 ms         1.68 ms          416 bytes_per_second=338.6M/s
repeat_strings_column_times/16384/16/manual_time        0.143 ms        0.158 ms         4756 bytes_per_second=1.34724G/s
repeat_strings_column_times/16384/64/manual_time        0.413 ms        0.430 ms         1676 bytes_per_second=1.44544G/s
repeat_strings_column_times/16384/256/manual_time        1.68 ms         1.70 ms          412 bytes_per_second=1.30622G/s
repeat_strings_column_times/65536/16/manual_time        0.414 ms        0.432 ms         1683 bytes_per_second=1.86912G/s
repeat_strings_column_times/65536/64/manual_time         2.54 ms         2.56 ms          274 bytes_per_second=959.96M/s
repeat_strings_column_times/65536/256/manual_time        32.9 ms         32.9 ms           21 bytes_per_second=273.805M/s
repeat_strings_column_times/262144/16/manual_time        3.67 ms         3.69 ms          190 bytes_per_second=864.283M/s
repeat_strings_column_times/262144/64/manual_time        46.0 ms         46.0 ms           15 bytes_per_second=212.259M/s
repeat_strings_column_times/262144/256/manual_time        329 ms          329 ms            2 bytes_per_second=109.717M/s
```

### Performance graph:

image

image

### CODE REVIEW

**Comment by ttnghia** [comment]
<details><summary>Plot Data</summary>
Without any overflow checking: | With separate overflow checking: | With integrated overflow checking: |   | Slowdown by separate checking | Slowdown by integrated checking
-- | -- | -- | -- | -- | --
45 | 37 | 42 |   | 0.8222222222 | 0.9333333333
156 | 127 | 134 |   | 0.8141025641 | 0.858974359
296 | 267 | 265 |   | 0.902027027 | 0.8952702703
212 | 165 | 143 |   | 0.7783018868 | 0.6745283019
636 | 511 | 525 |   | 0.8034591195 | 0.8254716981
1079 | 834 | 896 |   | 0.7729379055 | 0.8303985171
808 | 542 | 554 |   | 0.6707920792 | 0.6856435644
1435 | 1168 | 1127 |   | 0.8139372822 | 0.7853658537
1147 | 1075 | 1104 |   | 0.9372275501 | 0.962510898
2326 | 1823 | 1806 |   | 0.7837489252 | 0.7764402408
2212 | 2060 | 2020 |   | 0.931283906 | 0.9132007233
290 | 230 | 261 |   | 0.7931034483 | 0.9
4844 | 4224 | 3862 |   | 0.8720066061 | 0.7972749794
2622 | 2471 | 2590 |   | 0.9424103738 | 0.9877955759
238 | 233 | 240 |   | 0.9789915966 | 1.008403361
6175 | 6367 | 6447 |   | 1.031093117 | 1.044048583
2737 | 2730 | 2741 |   | 0.9974424552 | 1.001461454
245 | 244 | 245 |   | 0.9959183673 | 1
43 | 31 | 38 |   | 0.7209302326 | 0.8837209302
45 | 41 | 43 |   | 0.9111111111 | 0.9555555556
43 | 42 | 42 |   | 0.976744186 | 0.976744186
100 | 85 | 96 |   | 0.85 | 0.96
98 | 93 | 97 |   | 0.9489795918 | 0.9897959184
85 | 83 | 85 |   | 0.9764705882 | 1
394 | 314 | 369 |   | 0.7969543147 | 0.9365482234
380 | 360 | 373 |   | 0.9473684211 | 0.9815789474
339 | 337 | 338 |   | 0.994100295 | 0.9970501475
1431 | 1231 | 1347 |   | 0.8602375961 | 0.9412997904
1471 | 1395 | 1445 |   | 0.9483344663 | 0.982324949
1313 | 1293 | 1306 |   | 0.9847677075 | 0.9946686976
1866 | 1844 | 1869 |   | 0.988210075 | 1.001607717
972 | 905 | 959 |   | 0.9310699588 | 0.9866255144
274 | 270 | 273 |   | 0.9854014599 | 0.996350365
839 | 764 | 864 |   | 0.9106078665 | 1.029797378
211 | 209 | 212 |   | 0.990521327 | 1.004739336
109 | 109 | 109 |   | 1 | 1


</details>

**Comment by jlowe** [comment]
Here's the time comparisons for integrated checking vs no checking:
```
Benchmark                                                     Time        CPU Time
----------------------------------------------------------------------------------
repeat_strings_scalar_times/256/16/manual_time            0.003 ms        0.002 ms
repeat_strings_scalar_times/256/64/manual_time            0.008 ms        0.008 ms
repeat_strings_scalar_times/256/256/manual_time           0.014 ms        0.014 ms
repeat_strings_scalar_times/1024/16/manual_time           0.019 ms         0.02 ms
repeat_strings_scalar_times/1024/64/manual_time           0.011 ms        0.012 ms
repeat_strings_scalar_times/1024/256/manual_time          0.025 ms         0.03 ms
repeat_strings_scalar_times/4096/16/manual_time           0.019 ms        0.021 ms
repeat_strings_scalar_times/4096/64/manual_time           0.025 ms        0.026 ms
repeat_strings_scalar_times/4096/256/manual_time           0.03 ms         0.03 ms
repeat_strings_scalar_times/16384/16/manual_time          0.016 ms        0.018 ms
repeat_strings_scalar_times/16384/64/manual_time          0.023 ms        0.026 ms
repeat_strings_scalar_times/16384/256/manual_time          0.84 ms         0.84 ms
repeat_strings_scalar_times/65536/16/manual_time          0.028 ms        0.032 ms
repeat_strings_scalar_times/65536/64/manual_time          0.011 ms        0.011 ms
repeat_strings_scalar_times/65536/256/manual_time          -0.3 ms         -0.3 ms
repeat_strings_scalar_times/262144/16/manual_time        -0.015 ms       -0.016 ms
repeat_strings_scalar_times/262144/64/manual_time           0.0 ms          0.0 ms
repeat_strings_scalar_times/262144/256/manual_time          0.0 ms          0.0 ms
repeat_strings_column_times/256/16/manual_time            0.008 ms        0.009 ms
repeat_strings_column_times/256/64/manual_time            0.008 ms        0.009 ms
repeat_strings_column_times/256/256/manual_time           0.011 ms        0.011 ms
repeat_strings_column_times/1024/16/manual_time           0.005 ms        0.005 ms
repeat_strings_column_times/1024/64/manual_time           0.006 ms        0.006 ms
repeat_strings_column_times/1024/256/manual_time            0.0 ms         0.01 ms
repeat_strings_column_times/4096/16/manual_time           0.008 ms        0.008 ms
repeat_strings_column_times/4096/64/manual_time           0.008 ms        0.008 ms
repeat_strings_column_times/4096/256/manual_time            0.0 ms         0.01 ms
repeat_strings_column_times/16384/16/manual_time          0.008 ms        0.008 ms
repeat_strings_column_times/16384/64/manual_time          0.007 ms        0.007 ms
repeat_strings_column_times/16384/256/manual_time          0.01 ms         0.01 ms
repeat_strings_column_times/65536/16/manual_time         -0.001 ms       -0.001 ms
repeat_strings_column_times/65536/64/manual_time           0.03 ms         0.03 ms
repeat_strings_column_times/65536/256/manual_time           0.1 ms          0.1 ms
repeat_strings_column_times/262144/16/manual_time         -0.11 ms        -0.11 ms
repeat_strings_column_times/262144/64/manual_time          -0.1 ms         -0.1 ms
repeat_strings_column_times/262144/256/manual_time          0.0 ms          0.0 ms

```

These small time differences seem reasonable to me to have the check be integrated, but I'm curious if others agree the check is worth the additional cost.

**Comment by ttnghia** [comment]
I want to clarify the following:
 * Checking for overflow is a must for this API (`repeat_strings`)
 * Integrated checking, if implemented, will be turned off by default (as described in https://github.com/rapidsai/cudf/issues/8574) thus it will not affect anything else

**Comment by davidwendt** [comment]
I disagree with adding this check to the detail utility. The principle here is more than just about performance. 

We could split up `repeat_strings` into two APIs. One computes the output size the second materializes the strings column (given the precomputed output size as an optional argument). I have a somewhat generic design for this approach.

**Comment by harrism** [comment]
Agree. The fact that both approaches have the same performance supports separating the check, not integrating it. Also, if both require syncing, that is more support.

**Comment by ttnghia** [comment]
I just did a lot of cleanup for this PR. Now the code should look very clean for reviewing.

**Comment by ttnghia** [comment]
Rerun tests.

**Review by codereport — APPROVED** [review]
lgtm :+1:

**Review by robertmaynard — APPROVED** [review]

File: cpp/benchmarks/string/repeat_strings_benchmark.cpp [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+                          (strings_col.chars_size() + repeat_times_col.size() * sizeof(int32_t)));
+}
+
+static void generate_bench_args(benchmark::internal::Benchmark* b)
+{
+  int const min_rows   = 1 << 8;
+  int const max_rows   = 1 << 18;
+  int const row_mult   = 4;
+  int const min_rowlen = 1 << 4;
+  int const max_rowlen = 1 << 8;
```
Comment (vuule) [inline cpp/benchmarks/string/repeat_strings_benchmark.cpp:125]: should these be str_len?

Comment (ttnghia) [reply cpp/benchmarks/string/repeat_strings_benchmark.cpp:125]: Sure.

**Comment by harrism** [comment]
@ttnghia this status of this PR is conflicted. The target branch is still 21.08 but the project board says 21.10 (I made the latter change when it was not approved). Now that it is approved, if you are ready to merge you can go ahead and merge it into 21.08 as long as that is done before code freeze. But please move it to the 21.08 project board if you do so. Otherwise please change the target branch to 21.10 by clicking "edit" next to the title.

**Comment by ttnghia** [comment]
> @ttnghia this status of this PR is conflicted. The target branch is still 21.08 but the project board says 21.10 (I made the latter change when it was not approved). Now that it is approved, if you are ready to merge you can go ahead and merge it into 21.08 as long as that is done before code freeze. But please move it to the 21.08 project board if you do so. Otherwise please change the target branch to 21.10 by clicking "edit" next to the title.

Sorry for the confusion. I'm merging it to 21.08.

**Comment by ttnghia** [comment]
@gpucibot merge

**Pull request merged**